feat : Add support to customize the developer account password (#2359)

rohanKanojia opened the pull request:
Fix #2539

**Relates to:** Issue #2539

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://github.com/crc-org/crc/blob/main/developing.adoc)
 - [x] My code follows the style guidelines of this project
 - [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [x] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [x] I have performed a self-review of my code
 - [x] I have added tests that prove my fix is effective or that my feature works
 - [x] New and existing unit tests pass locally with my changes
 - [x] I tested my code on specified platforms
   - [x] Linux
   - [ ] Windows
   - [ ] MacOS

## Solution/Idea

+ Generate developer password instead of hardcoded `developer` string as password
+ Add config option to set developer password for cluster `developer-password`

## Proposed changes
crc would no longer have `developer` as user password. It would be generated each time cluster is created (just like kubeadmin password). If user wants to override it, they can use `developer-password` configuration option in CRC config.

## Testing

- After starting CRC cluster user would see a random string as password value instead of `developer
- If user has specified `developer-password` configuration option in CRC config, then that value is used in developer password:
```sh
    $ crc config set developer-password mypassword
    $ crc start
    [...]
    INFO Adding crc-admin and crc-developer contexts to kubeconfig...
    Started the OpenShift cluster.
    
    The server is accessible via web console at:
      https://console-openshift-console.apps-crc.testing
    
    Log in as user:
      Username: developer
      Password: mypassword
```

rohanKanojia committed:
feat : Write developer password to file just like kubeadmin password (#2539)

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

anjannath commented:
/ok-to-test

praveenkumar commented:
looks like `prow/security` is not liking this https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-crc-org-crc-main-security (recent jobs are green but failed for this)

rohanKanojia commented:
@praveenkumar :  How can I reproduce this failure locally? I'm not able to get much idea about what's wrong by looking at logs:

```
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"sigs.k8s.io/prow/pkg/entrypoint/run.go:84","func":"sigs.k8s.io/prow/pkg/entrypoint.Options.internalRun","level":"error","msg":"Error executing test process","severity":"error","time":"2024-11-18T08:45:42Z"}
ERRO[2024-11-18T08:45:43Z] Some steps failed:                           
ERRO[2024-11-18T08:45:43Z] 
  * could not run steps: step security failed: "security" test steps failed: "security" pod "security-openshift-ci-security-snyk-scan" failed: could not watch pod: the pod ci-op-m0wwywcn/security-openshift-ci-security-snyk-scan failed after 44s (failed containers: test): ContainerFailed one or more containers exited
```

Could it be possible that it's an intermittent failure?

cfergeau commented:
The failure is
```
[Medium] Use of Hardcoded Credentials
   ID: 3373e961-2410-402e-8db9-9c0b595d4e62 
   Path: pkg/crc/config/settings.go, line 32 
   Info: Do not hardcode passwords in code. Found hardcoded saved in DeveloperPassword.
```
which corresponds to
```
	DeveloperPassword        = "developer-password"
```

I think @albfan did some work on snyk's false positives recently (?)

rohanKanojia commented:
@cfergeau: Thanks for checking. We discussed this in a crc internal Slack chat and decided to add an exception for this particular rule.

praveenkumar commented:
Without this PR the `developer` user have `developer` password and it doesn't have a way to change it. There are users who might be using this as scripted way and consuming this well known password. Now with this change either user before hand set the required password or they will get a random password, this looks like a UX change. I think what we should do is have default password set to be `developer` but provide option to user to change if they want this way the UX is not going to change and existing scripts/automation wouldn't fail. @crc-org/crc-team wdyt?

rohanKanojia commented:
@praveenkumar : Maybe we can add a configuration named `useLegacyDeveloperPassword` that would be true by default.

praveenkumar commented:
> `useLegacyDeveloperPassword`

@rohanKanojia I don't think it is good  to have multiple config option just for developer user.

cfergeau commented on pkg/crc/cluster/kubeadmin_password.go:
```diff
@@ -17,33 +17,21 @@ import (
+// GenerateUserPassword creates and put updated password to ~/.crc/machine/crc/ directory
+func GenerateUserPassword(passwordFile string, user string) error {
+	logging.Infof("Generating new password for the %s user", user)
 	kubeAdminPassword, err := GenerateRandomPasswordHash(23)
```
This could now be named `password` or such, this is no longer necessarily a kubeadmin password.

cfergeau commented on pkg/crc/cluster/kubeadmin_password.go:
```diff
@@ -17,33 +17,21 @@ import (
-
-	kubeAdminPassword, err := GetKubeadminPassword()
+// UpdateUserPassword updates the htpasswd secret
+func UpdateUserPassword(ctx context.Context, ocConfig oc.Config, newKubeAdminPassword string, newDeveloperPassword string) error {
```
I'd use the plural here as there are multiple passwords involved `UpdateUserPasswords`

cfergeau commented on pkg/crc/cluster/kubeadmin_password.go:
```diff
@@ -17,33 +17,21 @@ import (
-	kubeAdminPassword, err := GetKubeadminPassword()
+// UpdateUserPassword updates the htpasswd secret
+func UpdateUserPassword(ctx context.Context, ocConfig oc.Config, newKubeAdminPassword string, newDeveloperPassword string) error {
+	credentials, err := resolveUserPassword(newKubeAdminPassword, newDeveloperPassword, constants.GetKubeAdminPasswordPath(), constants.GetDeveloperPasswordPath())
```
Same here, `resolveUserPasswords`

cfergeau commented on pkg/crc/cluster/kubeadmin_password.go:
```diff
@@ -50,7 +50,7 @@ func UpdateUserPassword(ctx context.Context, ocConfig oc.Config, newKubeAdminPas
 
-	logging.Infof("Changing the password for the kubeadmin user")
+	logging.Infof("Changing the password for the users")
 	expected, err := getHtpasswd(credentials, externals)
```
Did this belong in the previous commit?

cfergeau commented on pkg/crc/cluster/kubeadmin_password.go:
```diff
@@ -60,7 +60,7 @@ func UpdateUserPassword(ctx context.Context, ocConfig oc.Config, newKubeAdminPas
 	_, stderr, err = ocConfig.RunOcCommandPrivate(cmdArgs...)
 	if err != nil {
-		return fmt.Errorf("Failed to update kubeadmin password %v: %s", err, stderr)
+		return fmt.Errorf("failed to update user passwords %v: %s", err, stderr)
```
Same question?

cfergeau commented on pkg/crc/config/settings.go:
```diff
@@ -134,6 +135,8 @@ func RegisterSettings(cfg *Config) {
 
 	cfg.AddSetting(KubeAdminPassword, "", validateString, SuccessfullyApplied,
 		"User defined kubeadmin password")
+	cfg.AddSetting(DeveloperPassword, "", validateString, SuccessfullyApplied,
```
What if we use `"developer"` here as the default value (second parameter to `AddSetting`), wouldn't this make the second commit redundant?

rohanKanojia replied:
This change was added in second commit: e9a11fc31e65e4e8d88beacac7effddba84ca152

Did you mean we can avoid third commit by adding developer as default value?

cfergeau replied:
I don't remember exactly what I meant, but the current iteration has `cfg.AddSetting(DeveloperPassword, constants.DefaultDeveloperPassword, ...` which is what I was hoping/suggesting we could use, so all is well :)

rohanKanojia committed:
feat : Add config option to set developer password for cluster (#2539)

Add option to set developer password for the user.

```
$ crc config set developer-password mypassword
$ crc start
[...]
INFO Adding crc-admin and crc-developer contexts to kubeconfig...
Started the OpenShift cluster.

The server is accessible via web console at:
  https://console-openshift-console.apps-crc.testing

Log in as user:
  Username: developer
  Password: mypassword
```

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

gbraad commented:
Can we add an integration test to verify this works as expected?

rohanKanojia commented:
@gbraad: Will add an e2e test to verify these scenarios:
- **Given** user has no developer-password configured, **when** user performs crc start, **then** default developer password is `developer`
- **Given** user has configured developer-password as `secret`,  **when** user performs crc start, **then** developer password is `secret`

rohanKanojia committed:
test (e2e) : add e2e test for developer-password config option

+ Add additional assertions in basic test to verify that when no developer password is configured,
  then "developer" value is used for password
+ Add custom_developer_password e2e test to verify that when custom developer-password configuration
  option is provided, then that custom value is used as developer password

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

cfergeau reviewed (approved):
/lgtm
/approve
